Gradle custom java zippublish plugin

prudhvigodithi opened the pull request:
### Description
This change will add custom zippublish plugin, that can publish plugin zips to maven repo.
 
### Issues Resolved
https://github.com/opensearch-project/opensearch-build/issues/1916
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

peterzhuamazon commented on buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java:
```diff
@@ -120,23 +120,25 @@ public String call() throws Exception {
-            if (project.getPluginManager().hasPlugin("opensearch.java")) {
-                publication.artifact(project.getTasks().getByName("sourcesJar"));
-                publication.artifact(project.getTasks().getByName("javadocJar"));
+            String search  = "zip";
```
Hi @prudhvigodithi,

A bit confused.

Can we have a better naming for this var?
`search` does not tell me anything on it.

And the logic seems to find whether these is any zips, before attempting the publish of java later.

Thanks.

prudhvigodithi replied:
Hi @peterzhuamazon 
Fixed access modifiers and var strings in my latest commit.

dblock commented:
Will gladly review once you think this is ready!

prudhvigodithi commented:
Thanks @dblock , 
the reason I still left as draft is, I still dont see  the plugin gets added to local maven repo (using build tools) during runtime so that downstream components can use it while the plugin is applied as` apply plugin: 'opensearch.zippublish'`
However when I add the plugin manually to local maven repo then it gets picked when apply plugin is added.
So just looking for some suggestions if anything I have missed. Just to summarize the plugin functionality works fine as expected, just the pending part is how to integrate with the existing custom plugin setup, so that even this is  added to local maven repo during runtime and each component can then use it.
@saratvemulapalli

dblock commented on buildSrc/src/main/groovy/org/opensearch/gradle/test/TestWithSslPlugin.java:
```diff
@@ -49,6 +49,7 @@
 public class TestWithSslPlugin implements Plugin<Project> {
 
     @Override
+    @SuppressWarnings({ "unchecked" })
```
Why this? Feels unrelated, move into a separate PR.

prudhvigodithi replied:
This should fix the warning 
`Note: /Users/pgodithi/git/OpenSearch/buildSrc/src/main/groovy/org/opensearch/gradle/test/TestWithSslPlugin.java uses unchecked or unsafe operations.`
I will add it to new PR later.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java:
```diff
@@ -120,23 +120,25 @@ public String call() throws Exception {
-                publication.artifact(project.getTasks().getByName("javadocJar"));
+            // To exclude java artifcats for maven zip type publications
+            String zipPublicationSearch = "zip";
+            if (!publication.getName().toLowerCase().contains(zipPublicationSearch.toLowerCase())) {
```
Is this necessary? It looks like you're creating a second kind of publication (called `zippublish`) that is different from existing publication (called `publish`). Which is fine, but I would expect the publish plugin to not be touched/modified at all, or see Zips be published the same way as `sourcesJar` and `javadocJar` with a 1-liner of `publication.artifact(project.getTasks().getByName("....Zip"));`.

prudhvigodithi replied:
I dont need to do zipPublicationSearch check anymore as I'm suspending the tasks in the ZipPublish.java. Please check my latest commit.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,146 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
```
New files only need SPDX header, they are not licensed to Elasticsearch. Remove this and the modifications below.

dblock reviewed (changes requested):
Something looks odd about introducing a second kind of publication while that wasn't the case for bin or source JARs.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java:
```diff
@@ -132,11 +131,10 @@ public String call() throws Exception {
-
-            generatePomTask.configure(
-                t -> t.dependsOn(String.format("generatePomFileFor%sPublication", Util.capitalize(publication.getName())))
-            );
```
Why this change? Feels like the dependency should always be there regardless of what's included in the artifacts.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,122 @@
+                    maven.setUrl(buildDirectory.toString() + LOCAL_STAGING_REPO_PATH);
+                });
+            });
+            System.out.println("Starting " + MAVEN_ZIP_PUBLISH_TASK + " task");
```
Remove `System.out.println`.

prudhvigodithi replied:
Will be removing in my next commit.

prudhvigodithi replied:
Just removed. it and also cleaned up the code   
 
 ```
public final static String MAVEN_ZIP_PUBLISH_TASK = "publish"
        + ZipPublishUtil.capitalize(PUBLICATION_NAME)
        + "PublicationTo"
        + ZipPublishUtil.capitalize(STAGING_REPO)
        + "Repository";

```

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,122 @@
+                            System.out.println("");
+                        }
+                    }
+                    String finalZipVersion = version + extraSuffix;
```
This is backwards. You have access to version and qualifier properties, so the last thing you should be doing is parsing the ZIP filename. Build the correct version from those.

prudhvigodithi replied:
Hey @dblock, the `$VERSION` currently already has the qualifier and snapshot info ex `2.0.0-rc1-SNAPSHOT`, it will save some flags for running the task and would reduce the human errors
Now a user just need to pass `./gradlew publishMavenzipPublicationToZipstagingRepository -Dopensearch.version=$VERSION` , dont need to pass any additional flags.

prudhvigodithi replied:
The version that is getting passed has the value as `VERSION=2.0.0-alpha1-SNAPSHOT`, not just `2.0.0`, hence I went with that parsing logic as in `$VERSION` itself I can get the qualifier and snapshot info. 
so now a user need not pass all other flags `version_qualifier` and `build.snapshot`. I can try with `VersionProperties.getOpenSearch()`, but again it this only gives just the version number, then we expect a user to add the other flags to execute the task.

prudhvigodithi replied:
from the existing task `bundlePlugin` coming from [PluginBuildPlugin](https://github.com/opensearch-project/OpenSearch/blob/05a58192433f6e007ade61b4b4f21d0f068d50e0/buildSrc/src/main/java/org/opensearch/gradle/plugin/PluginPropertiesExtension.java), the output is an generated as a zip file, I have added the publish artifact to look for this output.

prudhvigodithi replied:
The `zipArtifact` name will be the project root name and `version` is the gradle property of the project.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,122 @@
+                    // -PzipFilePath=/build/distributions/opensearch-job-scheduler-2.0.0.0-alpha1-SNAPSHOT.zip
+                    if (getProperty("zipFilePath") != null) {
+                        BUILD_DISTRIBUTIONS_LOCATION = getProperty("zipFilePath");
+                        zipFilePath = BUILD_DISTRIBUTIONS_LOCATION + zipArtifact + "-" + finalZipVersion + ".zip";
```
This logic is the same, just `BUILD_DISTRIBUTIONS_LOCATION` changes.

All these CAPITALIZED variables are weird. Only CONSTANTS should be capitalized and I suspect half of them aren't needed anyway. Cleanup.

prudhvigodithi replied:
I have considered them as constants (using variable modifiers static and final ) :) 
As these vars are constants, and have added final to them, these referenced one or multiple times throughout the code, also these should not be changed moving forward with any logic, hence considering these I have CAPITALIZED them.  Please correct me if my though process is wrong, tomorrow if we have to change some name setting, staging repo settings and distributions path, all we need is to just update the constants, no need to touch the actual methods.

prudhvigodithi replied:
Agree with `BUILD_DISTRIBUTIONS_LOCATION`, also when I read back again, better it would be as `zipDistributionLocation`, fixed this in my latest commit.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublishExtension.java:
```diff
@@ -0,0 +1,22 @@
+        this.zipGroup = zipGroup;
+    }
+
+    public String getZipgroup() {
```
getZipGroup

prudhvigodithi replied:
Will push this change in my next commit.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublishUtil.java:
```diff
@@ -0,0 +1,16 @@
+        return str.substring(0, 1).toUpperCase() + str.substring(1);
+    }
+
+}
```
I think you call this with well known values, so just hard-code those values and remove this method.

prudhvigodithi replied:
Util's I have added in consideration for future as well, we might to merge the existing publish task and this zippublish plugin to one common plugin, so then these handy Util's will come into picture. :)

dblock replied:
1. Publication name is a constant, `public final static String PUBLICATION_NAME = "mavenzip";`.
2. `MAVEN_ZIP_PUBLISH_POM_TASK` is a constant that uses that, `public final static String MAVEN_ZIP_PUBLISH_POM_TASK = "generatePomFileFor" + ZipPublishUtil.capitalize(PUBLICATION_NAME) + "Publication";`
3. So nothing is dynamic that needs to call `capitalize`, just do

```java
public final static String PUBLICATION_NAME = "mavenzip";
public final static String MAVEN_ZIP_PUBLISH_POM_TASK = "generatePomFileForMavenZipPublication";
```

prudhvigodithi replied:
off course I can hardcode them, but if we change the `PUBLICATION_NAME`, then its expected from user to also modify `MAVEN_ZIP_PUBLISH_POM_TASK` to fix the publication, hence I added the Util.

dblock replied:
Why would we ever want to change these constants? I think you're over-optimizing for a complicated future problem by adding a new method (and no tests for it).

dblock commented on buildSrc/src/test/groovy/org/opensearch/gradle/zipplugin/ZipPublishTest.groovy:
```diff
@@ -0,0 +1,21 @@
+
+
+class ZipPublishTest {
+    public void zippublish_plugin_should_add_task_to_project() {
```
I think the convention here is `testNamesMustBeCamelCase`?

You also need to actually implement tests.

prudhvigodithi replied:
Good point, will make this change.

prudhvigodithi commented:
@dblock checking to add some more unit tests, as this not an actual task (tasks unit tests could be just calling the task), but a apply() method, I might need to add some dummy task that can use this apply() method.

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,82 @@
+    public final static String EXTENSION_NAME = "zipmavensettings";
+    public final static String PUBLICATION_NAME = "mavenzip";
+    public final static String STAGING_REPO = "zipstaging";
+    public final static String MAVEN_ZIP_PUBLISH_POM_TASK = "generatePomFileFor"
```
hardcode `generatePomFileForMavenZipPublication`, get rid of `ZipPublishUtil`

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,82 @@
+        Task javaDocJarTask = project.getTasks().findByName("javadocJar");
+        if (javaDocJarTask != null) {
+            javaDocJarTask.setEnabled(false);
+        }
```
This will conflict when multiple publication tasks are run. Care to explain what breaks if, for example, `sourceJarTask.setEnabled(false);` is not here?

prudhvigodithi replied:
I mean, anyhow I see we are not publishing the generated sourcesJar and javadocJar to actual nexus maven repo, this [publish](https://github.com/opensearch-project/opensearch-build/blob/main/publish/publish-snapshot.sh#L118-L119) code will ignore these jars before uploading them to nexus, so ideally these just exists on the local staging repo (filesystem) and not in maven repo.
We can also proceed to remove the task suspension from plugin code which will cause no harm but along with zips, javadocJar and sourcesJar are also included to local staging repo, and off course we stop them from publishing to nexus with the existing publish [code](https://github.com/opensearch-project/opensearch-build/blob/main/publish/publish-snapshot.sh#L118-L119) .
The other way is remove the task suspension and allow users to declare the suspension at their own will in their gradle project globally in settings .gradle as `startParameter.excludedTaskNames=["sourcesJar", "javadocJar"]` or from task using `./gradlew publishMavenzipPublicationToZipstagingRepository -Dopensearch.version=$VERSION -Dbuild.snapshot=$SNAPSHOT -Dbuild.version_qualifier=$QUALIFIER -x sourcesJar -x javadocJar`
So there are multiple options go with.

reta replied:
@prudhvigodithi 

> I mean, anyhow I see we are not publishing the generated sourcesJar and javadocJar to actual nexus maven repo ...

For plugins we do [1] (and for everything else as well), probably we only skip that for snapshots:

[mapper-size-1.3.1-javadoc.jar](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-javadoc.jar)                     2022-03-30 20:38    414035      
[mapper-size-1.3.1-javadoc.jar.asc](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-javadoc.jar.asc)                 2022-03-30 20:38       287      
[mapper-size-1.3.1-javadoc.jar.md5](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-javadoc.jar.md5)                 2022-03-30 20:38        32      
[mapper-size-1.3.1-javadoc.jar.sha1](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-javadoc.jar.sha1)                2022-03-30 20:38        40      
[mapper-size-1.3.1-javadoc.jar.sha256](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-javadoc.jar.sha256)              2022-03-30 20:38        64      
[mapper-size-1.3.1-javadoc.jar.sha512](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-javadoc.jar.sha512)              2022-03-30 20:38       128      
[mapper-size-1.3.1-sources.jar](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-sources.jar)                     2022-03-30 20:38      8245      
[mapper-size-1.3.1-sources.jar.asc](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-sources.jar.asc)                 2022-03-30 20:38       287      
[mapper-size-1.3.1-sources.jar.md5](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-sources.jar.md5)                 2022-03-30 20:38        32      
[mapper-size-1.3.1-sources.jar.sha1](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-sources.jar.sha1)                2022-03-30 20:38        40      
[mapper-size-1.3.1-sources.jar.sha256](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-sources.jar.sha256)              2022-03-30 20:38        64      
[mapper-size-1.3.1-sources.jar.sha512](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1-sources.jar.sha512)              2022-03-30 20:38       128      
[mapper-size-1.3.1.jar](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1.jar)                             2022-03-30 20:38     10416      
[mapper-size-1.3.1.jar.asc](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1.jar.asc)                         2022-03-30 20:38       287      
[mapper-size-1.3.1.jar.md5](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1.jar.md5)                         2022-03-30 20:38        32      
[mapper-size-1.3.1.jar.sha1](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1.jar.sha1)                        2022-03-30 20:38        40      
[mapper-size-1.3.1.jar.sha256](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1.jar.sha256)                      2022-03-30 20:38        64      
[mapper-size-1.3.1.jar.sha512](https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/mapper-size-1.3.1.jar.sha512)                      2022-03-30 20:38       128

We should not disable that.

[1] https://repo1.maven.org/maven2/org/opensearch/plugin/mapper-size/1.3.1/

prudhvigodithi replied:
Cool, thanks @reta, this info really helps, I can see them on releases repo as well example: https://aws.oss.sonatype.org/content/repositories/releases/org/opensearch/opensearch-job-scheduler/1.3.0.0/.

prudhvigodithi replied:
So based on this @reta and @dblock we should be going back to my previous changes to actually modify the existing PublishPlugin to avoid publication artifact's of sourcesJar and javadocJar if the publication is for zip.

To refer here are the changes we need to add to existing PublishPlugin, to make the sourcesJar and javadocJar are still enabled for other publications but not zip, changes are 

https://github.com/prudhvigodithi/OpenSearch/blob/gradleplugin-2.3/buildSrc/src/main/java/org/opensearch/gradle/PublishPlugin.java#L124-L145

prudhvigodithi replied:
I have made the above change for PublishPlugin and tested job-scheduler, anomaly-detection and alerting, looks good just published the zip and not the sourcesJar and javadocJar along with zips, but they are added for jar maven publication, so no interruption for existing jar publications.

dblock replied:
I don't see this change here, am I missing something?

dblock commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,82 @@
+                    // Getting the Zip group from created extension
+                    String zipGroup = extset.getZipGroup();
+                    String zipArtifact = project.getName();
+                    String zipVersion = ZipPublishUtil.getProperty("version", project);
```
The `zipVersion` property is required. Get rid of `ZipPublishUtil.getProperty` and just do `project.property()`. 

If you want a default, I believe you can do `project.findProperty('sampleNew') ?: 'default value'`.

prudhvigodithi replied:
https://github.com/opensearch-project/OpenSearch/pull/2988#discussion_r856592640
If thats the case I will remove the Util's, but having Util's will definitely ease to add more methods and reduces the modification of the core logic and will also reduce some hard codings.

dblock commented on buildSrc/src/test/groovy/org/opensearch/gradle/zipplugin/ZipPublishTest.groovy:
```diff
@@ -0,0 +1,22 @@
+
+        assertTrue(project.plugins.hasPlugin("opensearch.zippublish"))
+        assertTrue(project.tasks.publishMavenzipPublicationToZipstagingRepository instanceof ZipPublish)
+    }
```
You need to write tests that ensure that this new plugin actually works.

prudhvigodithi replied:
Added some more assertions to make sure plugin wiring works properly.

dblock reviewed (changes requested):
This looks a lot cleaner. My biggest remaining concerns are:

1. Disabling JAR generation when this plugin is used, which seems like it will conflict.
2. Plugin needs tests.

reta commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublishExtension.java:
```diff
@@ -0,0 +1,21 @@
+ */
+package org.opensearch.gradle.zipplugin;
+
+public class ZipPublishExtension {
```
I would replace 'zip` terminology with 'PluginDistribution` since this is reflects what exactly we intend to publish and where is it applicable, fe `PublishPluginDistributionExtension`

prudhvigodithi replied:
since i have organized as folder under `zippublish/` I guess it should be good, the entire effort of this plugin is to publish the zips, hence added with start `Zip` :)

reta replied:
I believe the notion of zip is implementation detail, what we are developing is publication mechanism for plugin distribution

dblock replied:
I am with @reta and would call this `PluginDistributionExtension`

reta commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,90 @@
+        project.getPluginManager().apply(MavenPublishPlugin.class);
+        project.getExtensions().configure(PublishingExtension.class, publishing -> {
+            publishing.repositories(repositories -> {
+                repositories.maven(maven -> {
```
Why do we need `LOCAL_STAGING_REPO_PATH`? I assume we should use appropriate Maven repository (local or not)

prudhvigodithi replied:
@reta `LOCAL_STAGING_REPO_PATH = "/build/local-staging-repo";`, this is a file system path, that we consider as a local staging repo, for maven (in terms on maven terminology) its just a publication repo, so considering this `local-staging-repo` which is added across build scripts I have added `LOCAL_STAGING_REPO_PATH`, which denotes that its a filesystem path for locally staging the artifacts.

reta replied:
I see, thanks @prudhvigodithi , not much we can do about it ...

reta commented on buildSrc/src/main/java/org/opensearch/gradle/zipplugin/ZipPublish.java:
```diff
@@ -0,0 +1,90 @@
+    private void configMaven(Project project) {
+        final Path buildDirectory = project.getRootDir().toPath();
+        project.getPluginManager().apply(MavenPublishPlugin.class);
+        project.getExtensions().configure(PublishingExtension.class, publishing -> {
```
Should we check that the project is actually `PluginBuildPlugin`?

prudhvigodithi replied:
@reta I have to plans add this in readme doc as prereq to make sure the zip is built with bundlePlugin task, since this zippublish logic is executed during plugin apply, the task will created and saved as `publishMavenzipPublicationToZipstagingRepository `, an error occurs only during the task runtime, when called in buildscript as `./gradlew publishMavenzipPublicationToZipstagingRepository -Dopensearch.version=$VERSION -Dbuild.snapshot=$SNAPSHOT -Dbuild.version_qualifier=$QUALIFIER`, so adding logic in code to check if the zip is being built or not with bundlePlugin task, could show as error log the entire project runtime and not at the task execution, as apply() methods are called first before all the tasks for the project are executed.

reta replied:
That's fine, thanks, as far as we could explain the error clearly, that this publishing task is only available for plugin distribution.

prudhvigodithi replied:
Ya I will also add the clear readme doc, that says this plugin publishes zip that is the output of bundlePlugin task `mavenZip.artifact(project.getTasks().named("bundlePlugin"));`

prudhvigodithi commented:
Hey @reta and @dblock I have done some code cleanup as per feedback and added the following.

- Removed the entire extension file, as this only used to fetch the zip's groupID, since its fixed as `org.opensearch.plugin`, I have hardcoded it, which makes the extensions file not required anymore.
- Edited the existing PublishPlugin, to skip if the publication has name `zip` in it.
- Added proper unit tests, that covers edge cased related to maven zip publish.
- Tested with job-scheduler, alerting and anomaly-detection plugin, able to get the zips published excluding the sourceJar and javadocJar only for this zip publication.

@bbarani

dblock reviewed (commented):
I am good with these changes if @reta is good with this. I don't see any problems. Good job on the tests, too.

I do want to discuss naming one more time, once we merge this it will be difficult to rename all over the place. The proposed `publishMavenzipPublicationToZipstagingRepository` does not match the current pattern. 

These are the current publish tasks:

```
publish - Publishes all publications produced by this project.
publishAllPublicationsToSnapshotsRepository - Publishes all Maven publications produced by this project to the Snapshots repository.
publishAllPublicationsToTestRepository - Publishes all Maven publications produced by this project to the test repository.
publishNebulaPublicationToMavenLocal - Publishes Maven publication 'nebula' to the local Maven repository.
publishNebulaPublicationToSnapshotsRepository - Publishes Maven publication 'nebula' to Maven repository 'Snapshots'.
publishNebulaPublicationToTestRepository - Publishes Maven publication 'nebula' to Maven repository 'test'.
publishPluginMavenPublicationToMavenLocal - Publishes Maven publication 'pluginMaven' to the local Maven repository.
publishPluginMavenPublicationToSnapshotsRepository - Publishes Maven publication 'pluginMaven' to Maven repository 'Snapshots'.
publishPluginMavenPublicationToTestRepository - Publishes Maven publication 'pluginMaven' to Maven repository 'test'.
publishShadowPublicationToMavenLocal - Publishes Maven publication 'shadow' to the local Maven repository.
publishShadowPublicationToSnapshotsRepository - Publishes Maven publication 'shadow' to Maven repository 'Snapshots'.
publishShadowPublicationToTestRepository - Publishes Maven publication 'shadow' to Maven repository 'test'.
publishToMavenLocal - Publishes all Maven publications produced by this project to the local Maven cache.
```

I think what would make sense given this pattern is `publishPluginZipPublicationTo*` or `publishZipPublicationTo*`, because:

1.  We are not publishing a maven zip, we're publishing the _plugin_ zip.
2. CamelCase should match.

Similarly, the name of the plugin `org.opensearch.gradle.zipplugin.ZipPublish` would then be `org.opensearch.gradle.pluginzip.Publish`, no need for a double zip.

@prudhvigodithi WDYT? Possible?

prudhvigodithi commented:
> @prudhvigodithi WDYT? Possible?

yes @dblock possible, but how about the META-INF, is the existing one (`opensearch.zippublish`) fine ?, users should be able to use this with apply `apply plugin: 'opensearch.zippublish'`

dblock reviewed (approved):
LGTM, @reta?

dblock commented:
I didn't pay attention and this got merged onto 2.0, it shouldn't have been as 2.0 is stabilizing for RC1 😢

prudhvigodithi commented:
Created Manual Backport PR's
Backport to main: [PR](https://github.com/opensearch-project/OpenSearch/pull/3095)
Backport to 2.x: [PR](https://github.com/opensearch-project/OpenSearch/pull/3096)
